docs/proposals/28-autonomous-shoot-clusters.md

```diff
@@ -0,0 +1,497 @@
+
+<img src="../../logo/gardenadm-large.png" alt="gardenadm" width="100"/>
+
+In order to reduce cognitive load, we'll align the user experience of this new tool with what [`kubeadm`](https://kubernetes.io/docs/reference/setup-tools/kubeadm/) provides (note that `gardenadm` is not meant to replace `kubeadm` - we just want to reduce the entry barrier for people already familiar with `kubeadm`, [similar to how we have aligned the Gardener architecture with Kubernetes concepts](../../README.md#gardener)).
+After a cluster has been set up with `gardenadm`, the next step is to deploy a stripped-down version of [`gardenlet`](../concepts/gardenlet.md) into it.
+This `gardenlet` has coordinates for an existing Gardener system it can connect to such that it can register the `Shoot` resource (similar to how it registers the `Seed` resource today when running in a seed cluster).
+Any day-2 operations like updates or upgrades will be performed by the `gardenlet` and via the `Shoot` API.
+As a consequence, there will be no command in `gardenadm` similar to [`kubeadm upgrade`](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-upgrade/).
+It is explicitly out of scope to support cluster management without a `gardenlet` connected to a Gardener system (see [this section](#gardenadm-connect) for more details).
+
+### Prerequisites
+
+Similar to [`kubeadm`'s prerequisites](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#before-you-begin), `gardenadm` will assume that the following requirements are met for the machines:
+
+- One or more compatible Linux hosts.
+- Sufficient CPU and RAM (to be specified more concretely later).
+- Full network connectivity between all machines in the cluster (public or private network is fine).
+- [Certain ports](https://kubernetes.io/docs/reference/networking/ports-and-protocols/) must be open.
+- [systemd](https://www.freedesktop.org/wiki/Software/systemd/), [containerd](https://containerd.io/), and its [ctr (client CLI)](https://github.com/projectatomic/containerd/blob/master/docs/cli.md/) must be pre-installed.
+
+### Commands
+
+`gardenadm` will have the following commands:
+
+- [`gardenadm discover`](#gardenadm-discover) to discover the configuration of the Gardener system the autonomous shoot cluster shall be connected to later.
+- [`gardenadm init`](#gardenadm-init) to bootstrap a first control plane node.
+- [`gardenadm token`](#gardenadm-token) to manage bootstrap and discovery tokens for `gardenadm join`.
+- [`gardenadm join`](#gardenadm-join) to bootstrap further control plane nodes, or to bootstrap worker nodes and join them to the cluster.
+- [`gardenadm connect`](#gardenadm-connect) to deploy a `gardenlet` for further cluster management.
+- [`gardenadm version`](#gardenadm-version) to print the `gardenadm` version.
+
```

Comment:
Additionally, `gardenadm bootstrap` is mentioned later in the document, but not in this list.

Reply:
> We could also introduce a `gardenadm reset` command which reverts `gardenadm init` and `gardenadm join` on the node (similar to `kubeadm reset`).

I'd imagine that this can be challenging, especially, if there are already pods and containers deployed on a node. It's not as simple as removing GNA and kubelet units, or? Something (this command) has to stop the containers and revert whatever the kubelet has done. Maybe it's easier to not support it and rather wipe/redeploy the machine if needed?

> Additionally, `gardenadm bootstrap` is mentioned later in the document, but not in this list.

Yes, see https://github.com/gardener/gardener/pull/10536#discussion_r1799335045.

Reply:
In a `medium touch` scenario it might be the easiest thing to wipe the machine. 

More interesting use cases for `gardenadm reset` could be in the `high touch` scenario. In this case the `machine` could be a physical machine too or VMs where provisioning and de-provisioning require some manual effort.
These are the use cases I have in mind:
- The user did something wrong and want's to start all over again
- The user replaces one machine with another one and decommission the first one
  - They could add a new machine via `gardenadm join`.
  - Then the old machine could be reset via `gardenadm reset`
  - Finally, the old machine should be decommissioned and all configs, tokens and local disks created by Gardener should be delete, that it is safe to return the machine.

`gardenadm reset` would be probably complex. I cannot judge right now, if it is worth the effort.
Anyway, these are the steps which could be performed.
- Remove the own etcd from the etcd-cluster
- Stop all static pods running on the machine
- Stop and delete `gardener-node-agent` and `kubelet` systemd units
- Stop all containers started by `kubelet` in containerd
- Delete local etcd disk and Gardener configs (config files, token etc.)

Reply:
I agree that there might be use-cases, and I also agree that implementing such a command would be complex. It's not only that it would need to revert the things that it has performed itself on the machine, but also all the things that have been done by other (potentially third-party) components like `kubelet`, or extensions. This means that it needs to incorporate knowledge of what these components might have done.

Hence, I'd prefer to not add it to the scope for now and rather ask the user to wipe the disk/machine somehow (they probably know what needs to be done) if they want to get rid of the node.